### PR TITLE
Add e2e test for kubedns to forward PTR records to upstream server.

### PR DESCRIPTION
**Pull request by grayluck**
**What this PR does / why we need it**:
Add e2e test for kubedns to forward PTR records to upstream server.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/network/dns.go [hunk 1]
```diff
@@ -318,4 +318,8 @@ var _ = SIGDescribe("DNS", func() {
 
 		// TODO: Add more test cases for other DNSPolicies.
 	})
+
+	It("should forward PTR records to upstream server by kubedns", func() {
+
+	})
```
Comment (krmayankk) [inline test/e2e/network/dns.go]: where is the actual test ;-)

Comment (grayluck) [reply test/e2e/network/dns.go]: This PR is not ready for review yet. It's a place holder.
I will work on this when the kubedns side is merged ([PR#220 in k8s/dns](https://github.com/kubernetes/dns/pull/220)) and will remove [WIP] when this one is ready. Sorry for confusion.

**Comment by grayluck** [comment]
Please do not test right now. It's not passing until dns image is updated.

**Comment by grayluck** [comment]
/assign @MrHohn 
No hurry let's review it next week.

File: test/e2e/network/dns_configmap.go [hunk 2]
```diff
@@ -193,3 +193,46 @@ var _ = SIGDescribe("DNS configMap nameserver", func() {
...
+	t.init()
+
+	t.createUtilPod()
+	defer t.deleteUtilPod()
+
+	t.createDNSServerWithPtrRecord()
+	defer t.deleteDNSServerPod()
+
+	t.setConfigMap(&v1.ConfigMap{Data: map[string]string{
+		"stubDomains":         fmt.Sprintf(`{"acme.local":["%v"]}`, t.dnsServerPod.Status.PodIP),
```
Comment (MrHohn) [inline test/e2e/network/dns_configmap.go]: Seems like this stubdomain setup is not needed?

File: test/e2e/network/dns_common.go [hunk 3]
```diff
@@ -280,6 +283,38 @@ func (t *dnsTestCommon) createDNSServer(aRecords map[string]string) {
...
+			GenerateName: "e2e-dns-configmap-dns-server-",
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "dns",
+					Image: imageutils.GetE2EImage(imageutils.DNSMasq),
+					Command: []string{
+						"/bin/sh", "-c",
+						"echo \"192.0.2.123 my.test\" > /etc/dnsmasq-hosts && " +
```
Comment (MrHohn) [inline test/e2e/network/dns_common.go]: Is it possible to make the test PTR record an argument of the function? Seems easy to break if we need to keep them (here and the actual test) consistent.

File: test/e2e/network/dns_common.go [hunk 4]
```diff
@@ -280,6 +283,38 @@ func (t *dnsTestCommon) createDNSServer(aRecords map[string]string) {
...
+							"echo \"addn-hosts=/etc/dnsmasq-hosts\" >> /etc/dnsmasq.conf && " +
+							"/usr/sbin/dnsmasq -u root -k --log-facility - -q -A/my.test/1.2.3.4",
```
Comment (MrHohn) [inline test/e2e/network/dns_common.go]: Why do we need `-A/my.test/1.2.3.4`?

File: test/e2e/network/dns_common.go [hunk 5]
```diff
@@ -110,6 +110,9 @@ func (t *dnsTestCommon) runDig(dnsName, target string) []string {
 	default:
 		panic(fmt.Errorf("invalid target: " + target))
 	}
+	if strings.HasSuffix(dnsName, "in-addr.arpa") {
```
Comment (MrHohn) [inline test/e2e/network/dns_common.go]: minor: what if there is a trailing dot (e.g. `XXX.in-addr.arpa.`) ?

File: test/e2e/network/dns_configmap.go [hunk 6]
```diff
@@ -193,3 +193,46 @@ var _ = SIGDescribe("DNS configMap nameserver", func() {
...
+
+	t.c.CoreV1().ConfigMaps(t.ns).Delete(t.name, nil)
+	t.checkDNSRecordFrom(
+		"123.2.0.192.in-addr.arpa",
+		func(actual []string) bool { return len(actual) == 0 },
+		"dnsmasq",
+		moreForeverTestTimeout)
+}
+
+var _ = SIGDescribe("DNS configMap nameserver", func() {
```
Comment (MrHohn) [inline test/e2e/network/dns_configmap.go]: This seems to be in the same category as the other test (both testing `DNS configMap nameserver`), can we combine them?

**Review by MrHohn — COMMENTED** [review]
Thanks! Looks good overall.

File: test/e2e/network/dns_configmap.go (hunk 2 above)
Comment (grayluck) [reply test/e2e/network/dns_configmap.go]: Removed.

File: test/e2e/network/dns_common.go (hunk 3 above)
Comment (grayluck) [reply test/e2e/network/dns_common.go]: Found something from the manual. Fixed.

File: test/e2e/network/dns_common.go (hunk 4 above)
Comment (grayluck) [reply test/e2e/network/dns_common.go]: No we don't.

File: test/e2e/network/dns_common.go (hunk 5 above)
Comment (grayluck) [reply test/e2e/network/dns_common.go]: Done.

File: test/e2e/network/dns_configmap.go (hunk 6 above)
Comment (grayluck) [reply test/e2e/network/dns_configmap.go]: Combined.

**Review by MrHohn — APPROVED** [review]
/lgtm

**Comment by grayluck** [comment]
Thanks for review!

File: test/e2e/network/dns_configmap.go [hunk 7]
```diff
@@ -229,10 +219,20 @@ func (t *dnsPtrFwdTest) run() {
...
 var _ = SIGDescribe("DNS configMap nameserver", func() {
-	t := &dnsPtrFwdTest{dnsTestCommon: newDnsTestCommon()}
-	BeforeEach(func() { t.c = t.f.ClientSet })
+
+	nsTest := &dnsNameserverTest{dnsTestCommon: newDnsTestCommon()}
+	fwdTest := &dnsPtrFwdTest{dnsTestCommon: newDnsTestCommon()}
+
+	BeforeEach(func() {
+		nsTest.c = nsTest.f.ClientSet
+		fwdTest.c = fwdTest.f.ClientSet
```
Comment (MrHohn) [inline test/e2e/network/dns_configmap.go]: Actually, seems unnecessary to initialize both tests in BeforeEach, what about initialize them in each It block instead?

Comment (grayluck) [reply test/e2e/network/dns_configmap.go]: Done.

File: test/e2e/network/dns_configmap.go [hunk 8]
```diff
@@ -185,11 +185,51 @@ func (t *dnsNameserverTest) run() {
...
+		"dnsmasq",
+		moreForeverTestTimeout)
+}
+
 var _ = SIGDescribe("DNS configMap nameserver", func() {
-	t := &dnsNameserverTest{dnsTestCommon: newDnsTestCommon()}
-	BeforeEach(func() { t.c = t.f.ClientSet })
+
+	nsTest := &dnsNameserverTest{dnsTestCommon: newDnsTestCommon()}
+	fwdTest := &dnsPtrFwdTest{dnsTestCommon: newDnsTestCommon()}
```
Comment (MrHohn) [inline test/e2e/network/dns_configmap.go]: Why not move these into It block as well?

Comment (grayluck) [reply test/e2e/network/dns_configmap.go]: Done. The test above was also changed by the way.

**Pull request merged**